File: internal/content/shim/lua/darkmagic/gameplay/systems/movement.lua
-- Integrate authoritative velocity into bounded, collision-aware position.
--
-- Input does not run here. A fixed-tick command handler has already translated
-- player intent into velocity. This system only applies that simulation fact to
-- every entity with Position + Velocity + Bounds.

local ecs = require("dm.ecs/v1")

local M = {}

local function clamp(value, minimum, maximum)
    return math.max(minimum, math.min(maximum, value))
end

local function collision_cell(value)
    -- Positions describe subtile centers. Adding one half before flooring maps
    -- a centered footprint back onto the cell containing that center.
    return math.floor(value + 0.5)
end

local function footprint_blocked(collision, x, y, radius)
    if not collision then return false end
    for _, offset_x in ipairs({-radius, radius}) do
        for _, offset_y in ipairs({-radius, radius}) do
            if collision:blocked(collision_cell(x + offset_x), collision_cell(y + offset_y)) then
                return true
            end
        end
    end
    return false
end

local function move_x(position, velocity, bounds, collider, collision, elapsed)
    local x, y = position:get("x"), position:get("y")
    local radius = collider:get("radius")
    local candidate = clamp(x + velocity:get("x") * elapsed, radius, bounds:get("width") - radius)

    -- Collision cells are integers; continuous movement is rounded only at the
    -- query boundary so slow movement does not lose fractional progress.
    if not footprint_blocked(collision, candidate, y, radius) then
        position:set("x", candidate)
    end
end

local function move_y(position, velocity, bounds, collider, collision, elapsed)
    local x, y = position:get("x"), position:get("y")
    local radius = collider:get("radius")
    local candidate = clamp(y + velocity:get("y") * elapsed, radius, bounds:get("height") - radius)

    -- X is read again after the X step. Resolving the axes separately gives a
    -- simple sliding response when only one side of a diagonal is blocked.
    if not footprint_blocked(collision, x, candidate, radius) then
        position:set("y", candidate)
    end
end

local function update_entity(entity, collision, elapsed)
    local position = ecs.get(entity, "dm.world.position")
    local velocity = ecs.get(entity, "dm.world.velocity")
    local bounds = ecs.get(entity, "dm.world.bounds")
    local collider = ecs.get(entity, "dm.world.collider")
    move_x(position, velocity, bounds, collider, collision, elapsed)
    move_y(position, velocity, bounds, collider, collision, elapsed)
end

function M.register(collision)
    ecs.system({
        id = "darkmagic.world.integrate",
        phase = "movement",
        query = { all = { "dm.world.position", "dm.world.velocity", "dm.world.bounds", "dm.world.collider" } },
        read = { "dm.world.velocity", "dm.world.bounds", "dm.world.collider" },
        write = { "dm.world.position" },
        update = function(context, entities)
            for _, entity in ipairs(entities) do
                update_entity(entity, collision, context.delta_seconds)
            end
        end,
    })
end

return M

File: internal/game/world/movement.go
package world

import "math"

// IntegrateVelocity applies one bounded, axis-separated collision step. It is
// shared by authoritative Lua through the world-map capability and by client
// rollback/replay prediction.
func IntegrateVelocity(collision *Map, position, velocity, bounds Point, radius, elapsed float64) Point {
	result := position
	result.X = clampMovement(position.X+velocity.X*elapsed, radius, bounds.X-radius)
	if footprintWalkable(collision, result.X, result.Y, radius) {
		position.X = result.X
	}
	result = position
	result.Y = clampMovement(position.Y+velocity.Y*elapsed, radius, bounds.Y-radius)
	if footprintWalkable(collision, result.X, result.Y, radius) {
		position.Y = result.Y
	}
	return position
}

func footprintWalkable(collision *Map, x, y, radius float64) bool {
	if collision == nil {
		return true
	}
	reach := int(math.Ceil(radius))
	for offsetY := -reach; offsetY <= reach; offsetY++ {
		for offsetX := -reach; offsetX <= reach; offsetX++ {
			if math.Hypot(float64(offsetX), float64(offsetY)) > radius+0.5 {
				continue
			}
			flags, found := collision.FlagsAtPosition(x+float64(offsetX), y+float64(offsetY))
			if found && flags.Blocked() {
				return false
			}
		}
	}
	return true
}

func clampMovement(value, minimum, maximum float64) float64 {
	return math.Max(minimum, math.Min(maximum, value))
}
